### DISCUSSION TITLE
Correct usage of `assertQuerysetEqual`

### DISCUSSION THREAD

**Discussion by rc65** [sbdchd/django-types#142] [General]
Hey, thanks for your work on this repo. It's been very helpful!

I have a question about the best way to use `assertQuerysetEqual` in tests. I am following the Django tutorial [here](https://docs.djangoproject.com/en/4.1/intro/tutorial05/). It seems from the tutorial that I should be able to call `assertQuerysetEqual` with something of type `List[Question]`, however this is not supported by `django-types`.

I have looked at the [type stub](https://github.com/typeddjango/django-stubs/blob/6aecafc64f02f54857bfa59b4eaab013e09be745/django-stubs/test/testcases.pyi#L203-L210) and it seems to only allow lists with certain primitives or `Queryset[Any]`. Below is the stub I'm talking about:

```
    def assertQuerysetEqual(
        self,
        qs: Union[Iterator[Any], List[Model], QuerySet[Any], RawQuerySet[Any]],
        values: Union[
            List[None],
            List[Tuple[str, str]],
            List[date],
            List[int],
            List[str],
            Set[str],
            QuerySet[Any],
        ],
        transform: Union[Callable[..., Any], Type[str]] = ...,
        ordered: bool = ...,
        msg: Optional[str] = ...,
    ) -> None: ...
```

So, finally, my question is, what's the best way to use this function? In my test I've resorted to getting a `Queryset[Question]` object by doing the following:

```
    def test_past_question(self):
        """
        Questions with a pub_date in the past are displayed on the
        index page.
        """
        question = create_question(question_text="Past question.", days=-30)
        response = self.client.get(reverse("polls:index"))
        self.assertQuerysetEqual(
            response.context["latest_question_list"],
            Question.objects.filter(pk=question.id),
        )
```

but it would be nicer just to use a list containing the question object like they do in the tutorial, like this:

```
    def test_past_question(self):
        """
        Questions with a pub_date in the past are displayed on the
        index page.
        """
        question = create_question(question_text="Past question.", days=-30)
        response = self.client.get(reverse("polls:index"))
        self.assertQuerysetEqual(response.context["latest_question_list"], [question])
```

Am I using this the wrong way, or could a `List[Any]` type be added to the stub?

Many thanks,
Ross

**Comment by sbdchd** [comment]
truth be told, I've never seen / used `assertQuerysetEqual` before so my guess is the type stub is probably wrong and should be less strict

especially since it's used only testing so if it explodes at runtime, no big deal since it just fails the test


I think your fix would work!
